### PR TITLE
New scene manager GUI plugin that works with ign-gui's MinimalScene

### PR DESCRIPTION
**Pull request by chapulina**
# 🎉 New feature

Closes https://github.com/ignitionrobotics/ign-gui/issues/137

Requires https://github.com/ignitionrobotics/ign-gui/pull/221

## Summary

This is a proof of concept for how an `ign-gazebo` `GuiSystem` can work with `ign-gui`'s `MinimalScene`. The idea is to split the concerns:

* `ign-gui`'s `MinimalScene` instantiates a 3D scene and paints to the window
* `ign-gazebo`'s `GzSceneManager` adds, removes and updates visuals on that scene, based on Gazebo's entities and components.

The two plugins synchronize through the `Render` event, that's called by `MinimalScene` within the rendering thread.

This initial implementation supports adding entities to the scene and updating their poses. There's lots still to be figured out:

- [ ] Choose a render engine from the command line (this information needs to be sent to `ign-gui`'s `MinimalScene` before it creates a scene.)
- [ ] With the current implementation, `ign-gui`'s `MinimalScene` will always create a scene. That plugin needs to be updated so it can be attached to the sensor system's scene if that exists, when running on the same process (related to #556) PR https://github.com/ignitionrobotics/ign-gazebo/pull/793

Missing features that could possibly be moved to independent `ign-gui` plugins:

- [x] Markers https://github.com/ignitionrobotics/ign-gui/pull/247
- [x] Move to / follow https://github.com/ignitionrobotics/ign-gui/pull/226
- [x] Move camera https://github.com/ignitionrobotics/ign-gui/pull/231

Missing features that could possibly be moved to independent `ign-gazebo` plugins:

- [ ] Spawn
- [ ] Select https://github.com/ignitionrobotics/ign-gazebo/pull/854
- [ ] View collisions
- [ ] Transform controls https://github.com/ignitionrobotics/ign-gazebo/pull/854
- [ ] Record Video https://github.com/ignitionrobotics/ign-gazebo/pull/900

## Test it

 To try it out, run:

`ign gazebo --gui-config src/ign-gazebo/src/gui/gui.config rolling_shapes.sdf -v 4`

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

**Comment by ahcorde** [comment]
I was reviewing some of the plugin in ign-gazebo that you mentioned to move them to a independent plugin (inside ign-gazebo). In particular these 3:

 - Select
 - View collisions
 - Transform controls

All of them require the renderUtil class which means that we need to create a renderUtil object and in each one of them we should update it (I think these calls are very costly). 

Anyhow I will prototype something

File: src/gui/gui.config [hunk 1]
```diff
@@ -40,6 +40,20 @@
...
+      <line own="right" target="right"/>
+      <line own="top" target="top"/>
+    </anchors>
+    <property key="resizable" type="bool">false</property>
+    <property key="width" type="double">5</property>
+    <property key="height" type="double">5</property>
+    <property key="state" type="string">floating</property>
+    <property key="showTitleBar" type="bool">false</property>
+  </ignition-gui>
+</plugin>
```
Comment (ahcorde) [inline src/gui/gui.config:55]: When these two plugins are merged in ign-gui we should add them here

 - https://github.com/ignitionrobotics/ign-gui/pull/231
 - https://github.com/ignitionrobotics/ign-gui/pull/226

File: src/rendering/RenderUtil.cc [hunk 2]
```diff
@@ -1963,6 +1963,10 @@ void RenderUtilPrivate::RemoveRenderingEntities(
 /////////////////////////////////////////////////
 void RenderUtil::Init()
 {
+  // Already initialized
+  if (nullptr != this->dataPtr->scene)
```
Comment (ahcorde) [inline src/rendering/RenderUtil.cc:1967]:
```suggestion
  if (this->dataPtr->scene)
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by ahcorde** [comment]
- MarkerManager plugin: https://github.com/ignitionrobotics/ign-gui/pull/247

we should add to the `gui.config` file when the PR is merged

```xml
<plugin filename="MarkerManager" name="Marker manager">
  <ignition-gui>
    <title>Marker Manager</title>
    <property type="bool" key="showTitleBar">false</property>
    <property type="string" key="state">docked</property>
    <property key="resizable" type="bool">false</property>
    <property key="width" type="double">5</property>
    <property key="height" type="double">5</property>
    <property key="state" type="string">floating</property>
    <property key="showTitleBar" type="bool">false</property>
  </ignition-gui>
</plugin>
```

**Comment by ahcorde** [comment]
I was having a look to the `record video` plugin but there is a feature to record the video in lockstep mode which means I can create the plugin but instead of including the plugin in `ign-gui` it should live in `ign-gazebo`.

@chapulina, is this assumption correct ? or am i missing something ?

**Comment by chapulina** [comment]
> @chapulina, is this assumption correct ?

Good point, Ignition GUI doesn't have the concept of time, so it would take some extra efforts if we wanted to move that plugin there. I'm fine keeping it in Ignition Gazebo :+1:

File: src/rendering/RenderUtil.cc (hunk 2 above)
Comment (chapulina) [reply src/rendering/RenderUtil.cc:2005]: I don't think there's harm in being more explicit when checking for null pointers. Some classes may have custom comparison operators that don't return what we may expect. Do you think there's an advantage to leaving the `nullptr` out?

File: src/gui/gui.config (hunk 1 above)
Comment (chapulina) [reply src/gui/gui.config]: I removed the changes to `gui.config` for now and added an example world that uses the new plugins. This way, we can incrementally add functionality to the scene and only make the switch to it once it's ready.

Added these plugins to the new `minimal_scene.sdf` in fbceade

**Comment by chapulina** [comment]
@ahcorde , I'm opening this PR for review.

I reverted the changes to `gui.config`, so the current PR doesn't affect any current users.

The new scene manager can be tested with the `minimal_scene.sdf` example world.

One suggestion for testing is to:

* Start the example that uses this plugin: `ign gazebo minimal_scene.sdf`

File: examples/worlds/minimal_scene.sdf [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+           <line own="right" target="right"/>
+           <line own="top" target="top"/>
+         </anchors>
+         <property key="resizable" type="bool">false</property>
+         <property key="width" type="double">5</property>
+         <property key="height" type="double">5</property>
+         <property key="state" type="string">floating</property>
+         <property key="showTitleBar" type="bool">false</property>
+       </ignition-gui>
+     </plugin>
```
Comment (ahcorde) [inline examples/worlds/minimal_scene.sdf:99]: Indentation
```suggestion
       <!-- Plugins that add functionality to the scene -->
      <plugin filename="GzSceneManager" name="Scene Manager">
        <ignition-gui>
          <anchors target="3D View">
            <line own="right" target="right"/>
            <line own="top" target="top"/>
          </anchors>
          <property key="resizable" type="bool">false</property>
          <property key="width" type="double">5</property>
          <property key="height" type="double">5</property>
          <property key="state" type="string">floating</property>
          <property key="showTitleBar" type="bool">false</property>
        </ignition-gui>
      </plugin>
      <plugin filename="InteractiveViewControl" name="Interactive view control">
        <ignition-gui>
          <anchors target="3D View">
            <line own="right" target="right"/>
            <line own="top" target="top"/>
          </anchors>
          <property key="resizable" type="bool">false</property>
          <property key="width" type="double">5</property>
          <property key="height" type="double">5</property>
          <property key="state" type="string">floating</property>
          <property key="showTitleBar" type="bool">false</property>
        </ignition-gui>
      </plugin>
      <plugin filename="CameraTracking" name="Camera Tracking">
        <ignition-gui>
          <anchors target="3D View">
            <line own="right" target="right"/>
            <line own="top" target="top"/>
          </anchors>
          <property key="resizable" type="bool">false</property>
          <property key="width" type="double">5</property>
          <property key="height" type="double">5</property>
          <property key="state" type="string">floating</property>
          <property key="showTitleBar" type="bool">false</property>
        </ignition-gui>
      </plugin>
      <plugin filename="MarkerManager" name="Marker manager">
        <ignition-gui>
          <anchors target="3D View">
            <line own="right" target="right"/>
            <line own="top" target="top"/>
          </anchors>
          <property key="resizable" type="bool">false</property>
          <property key="width" type="double">5</property>
          <property key="height" type="double">5</property>
          <property key="state" type="string">floating</property>
          <property key="showTitleBar" type="bool">false</property>
        </ignition-gui>
      </plugin>
```

File: src/gui/plugins/scene_manager/GzSceneManager.cc [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+#include "GzSceneManager.hh"
+
+#include <ignition/common/Profiler.hh>
+#include <ignition/gui/Application.hh>
+#include <ignition/gui/GuiEvents.hh>
+#include <ignition/plugin/Register.hh>
+#include <ignition/rendering/RenderingIface.hh>
+#include <ignition/rendering/Scene.hh>
+#include <ignition/gui/MainWindow.hh>
```
Comment (ahcorde) [inline src/gui/plugins/scene_manager/GzSceneManager.cc:26]:
```suggestion
#include <ignition/common/Profiler.hh>
#include <ignition/gui/Application.hh>
#include <ignition/gui/GuiEvents.hh>
#include <ignition/gui/MainWindow.hh>
#include <ignition/plugin/Register.hh>
#include <ignition/rendering/RenderingIface.hh>
#include <ignition/rendering/Scene.hh>
```

**Review by ahcorde — APPROVED** [review]
Two minor style issues. Otherwise LGTM

**Pull request merged**